Cleanup switch when create fails

When create fails, it leaves an empty active switch behind:

```
$ opam compiler create ocaml:4.12 --switch ocomp412
...
# configure: error: ...
$ opam switch list
#  switch                           compiler                        description
...
→  ocomp412
```

--- a/lib/github_client.ml
+++ b/lib/github_client.ml
@@ -4,7 +4,8 @@ type pr_info = { source_branch : Branch.t; title : string }
 
 type t = { pr_info : Pull_request.t -> (pr_info, [ `Unknown ]) result }
 
-let pr_info t = t.pr_info
+let pr_info t pr =
+  (t.pr_info pr : (_, [ `Unknown ]) result :> (_, [> `Unknown ]) result)
 
 module Real = struct
   let pull_source_user branch =

--- a/lib/github_client.mli
+++ b/lib/github_client.mli
@@ -2,7 +2,7 @@ type pr_info = { source_branch : Branch.t; title : string }
 
 type t = { pr_info : Pull_request.t -> (pr_info, [ `Unknown ]) result }
 
-val pr_info : t -> Pull_request.t -> (pr_info, [ `Unknown ]) result
+val pr_info : t -> Pull_request.t -> (pr_info, [> `Unknown ]) result
 
 val real : t
 

--- a/lib/op.ml
+++ b/lib/op.ml
@@ -1,5 +1,11 @@
 open! Import
 
+let translate_error s =
+  let open Rresult.R in
+  reword_error (function
+    | `Unknown -> msgf "%s" s
+    | `Command_failed cmd -> msgf "%s - command failed: %a" s pp_cmd cmd)
+
 let create runner github_client source switch_name ~configure_command =
   let switch_name =
     match switch_name with
@@ -7,13 +13,15 @@ let create runner github_client source switch_name ~configure_command =
     | None -> Source.global_switch_name source
   in
   let description = Source.switch_description source github_client in
-  let open Rresult.R in
   let open Let_syntax.Result in
-  (let* () = Opam.create runner ~name:switch_name ~description in
-   let* url = Source.switch_target source github_client in
-   let* () = Opam.pin_add runner ~name:switch_name url ~configure_command in
-   Opam.set_base runner ~name:switch_name)
-  |> reword_error (fun `Unknown -> msgf "Cannot create switch")
+  (match Opam.create runner ~name:switch_name ~description with
+  | Ok () ->
+      let* url = Source.switch_target source github_client in
+      let* () = Opam.pin_add runner ~name:switch_name url ~configure_command in
+      Opam.set_base runner ~name:switch_name
+  | Error (`Command_failed _) -> Opam.remove_switch runner ~name:switch_name
+  | Error `Unknown -> Error `Unknown)
+  |> translate_error "Cannot create switch"
 
 type reinstall_mode = Quick | Full
 
@@ -22,8 +30,7 @@ let reinstall_packages_if_needed runner = function
   | Full -> Opam.reinstall_packages runner
 
 let reinstall runner mode ~configure_command =
-  let open Rresult.R in
   let open Let_syntax.Result in
   (let* () = Opam.reinstall_compiler runner ~configure_command in
    reinstall_packages_if_needed runner mode)
-  |> reword_error (fun `Unknown -> msgf "Could not reinstall")
+  |> translate_error "Could not reinstall"

--- a/lib/opam.ml
+++ b/lib/opam.ml
@@ -84,3 +84,6 @@ let reinstall_compiler runner ~configure_command =
 let reinstall_packages runner =
   run_opam runner
     [ A "reinstall"; A "--assume-built"; A "--working-dir"; ocaml_variants ]
+
+let remove_switch runner ~name =
+  run_opam runner [ A "switch"; A "remove"; A (Switch_name.to_string name) ]

--- a/lib/opam.mli
+++ b/lib/opam.mli
@@ -2,20 +2,34 @@ val create :
   Runner.t ->
   name:Switch_name.t ->
   description:string ->
-  (unit, [ `Unknown ]) result
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
 val pin_add :
   Runner.t ->
   name:Switch_name.t ->
   string ->
   configure_command:Bos.Cmd.t option ->
-  (unit, [ `Unknown ]) result
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
-val set_base : Runner.t -> name:Switch_name.t -> (unit, [ `Unknown ]) result
+val set_base :
+  Runner.t ->
+  name:Switch_name.t ->
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
-val update : Runner.t -> name:Switch_name.t -> (unit, [ `Unknown ]) result
+val update :
+  Runner.t ->
+  name:Switch_name.t ->
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
 val reinstall_compiler :
-  Runner.t -> configure_command:Bos.Cmd.t option -> (unit, [ `Unknown ]) result
+  Runner.t ->
+  configure_command:Bos.Cmd.t option ->
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
-val reinstall_packages : Runner.t -> (unit, [ `Unknown ]) result
+val reinstall_packages :
+  Runner.t -> (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result
+
+val remove_switch :
+  Runner.t ->
+  name:Switch_name.t ->
+  (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result

--- a/lib/runner.ml
+++ b/lib/runner.ml
@@ -1,14 +1,12 @@
 open! Import
 
+type run_error = [ `Command_failed of Bos.Cmd.t | `Unknown ]
+
 type t = {
   run :
-    ?extra_env:(string * string) list ->
-    Bos.Cmd.t ->
-    (unit, [ `Unknown ]) result;
+    ?extra_env:(string * string) list -> Bos.Cmd.t -> (unit, run_error) result;
   run_out :
-    ?extra_env:(string * string) list ->
-    Bos.Cmd.t ->
-    (string, [ `Unknown ]) result;
+    ?extra_env:(string * string) list -> Bos.Cmd.t -> (string, run_error) result;
 }
 
 module Real = struct
@@ -28,7 +26,7 @@ module Real = struct
     let open Let_syntax.Result in
     let* env = explicit_env extra_env in
     Bos.OS.Cmd.in_null |> Bos.OS.Cmd.run_in ?env cmd
-    |> Rresult.R.reword_error (fun _ -> `Unknown)
+    |> Rresult.R.reword_error (fun _ -> `Command_failed cmd)
 
   let run_out ?extra_env cmd =
     let open Let_syntax.Result in
@@ -56,6 +54,10 @@ let dry_run =
   let open Dry_run in
   { run; run_out }
 
-let run_out t = t.run_out
+let run_out t ?extra_env cmd =
+  (t.run_out ?extra_env cmd
+    : (_, run_error) result
+    :> (_, [> run_error ]) result)
 
-let run t = t.run
+let run t ?extra_env cmd =
+  (t.run ?extra_env cmd : (_, run_error) result :> (_, [> run_error ]) result)

--- a/lib/runner.mli
+++ b/lib/runner.mli
@@ -2,11 +2,11 @@ type t = {
   run :
     ?extra_env:(string * string) list ->
     Bos.Cmd.t ->
-    (unit, [ `Unknown ]) result;
+    (unit, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result;
   run_out :
     ?extra_env:(string * string) list ->
     Bos.Cmd.t ->
-    (string, [ `Unknown ]) result;
+    (string, [ `Command_failed of Bos.Cmd.t | `Unknown ]) result;
 }
 
 val real : t
@@ -17,10 +17,10 @@ val run :
   t ->
   ?extra_env:(string * string) list ->
   Bos.Cmd.t ->
-  (unit, [ `Unknown ]) result
+  (unit, [> `Command_failed of Bos.Cmd.t | `Unknown ]) result
 
 val run_out :
   t ->
   ?extra_env:(string * string) list ->
   Bos.Cmd.t ->
-  (string, [ `Unknown ]) result
+  (string, [> `Command_failed of Bos.Cmd.t | `Unknown ]) result

--- a/lib/source.mli
+++ b/lib/source.mli
@@ -1,6 +1,6 @@
 type t = Github_branch of Branch.t | Github_PR of Pull_request.t
 
-val switch_target : t -> Github_client.t -> (string, [ `Unknown ]) result
+val switch_target : t -> Github_client.t -> (string, [> `Unknown ]) result
 
 val switch_description : t -> Github_client.t -> string
 

--- a/test/unit/test_op.ml
+++ b/test/unit/test_op.ml
@@ -11,6 +11,8 @@ let run_mock loc expectations =
   let run ?extra_env cmd = run_mock (cmd, extra_env) in
   (run, check)
 
+let opam_cli_env = Some [ ("OPAMCLI", "2.0") ]
+
 let create_tests =
   let test name ?switch_name ?configure_command expectations ~expected =
     ( name,
@@ -29,16 +31,25 @@ let create_tests =
         Alcotest.check Alcotest.(result unit msg) __LOC__ expected got;
         check () )
   in
-  let create_call =
-    ( Bos.Cmd.(
-        v "opam" % "switch" % "create" % "USER-REPO-BRANCH" % "--empty"
-        % "--description" % "[opam-compiler] USER/REPO:BRANCH"),
-      Some [ ("OPAMCLI", "2.0") ] )
+  let create_cmd =
+    Bos.Cmd.(
+      v "opam" % "switch" % "create" % "USER-REPO-BRANCH" % "--empty"
+      % "--description" % "[opam-compiler] USER/REPO:BRANCH")
   in
+  let create_call = (create_cmd, opam_cli_env) in
   [
-    test "create: first create fails"
+    test "create: create fails with unknown error"
       [ Mock.expect create_call ~and_return:(Error `Unknown) ]
       ~expected:(Error (`Msg "Cannot create switch"));
+    test "create: when opam command fails"
+      [
+        Mock.expect create_call ~and_return:(Error (`Command_failed create_cmd));
+        Mock.expect
+          ( Bos.Cmd.(v "opam" % "switch" % "remove" % "USER-REPO-BRANCH"),
+            opam_cli_env )
+          ~and_return:(Ok ());
+      ]
+      ~expected:(Ok ());
   ]
 
 let reinstall_tests =
@@ -55,7 +66,6 @@ let reinstall_tests =
         Alcotest.check Alcotest.(result unit msg) __LOC__ expected got;
         check () )
   in
-  let opam_cli_env = Some [ ("OPAMCLI", "2.0") ] in
   [
     test "reinstall (quick)" Quick None
       Bos.Cmd.

